### PR TITLE
Introduce AArch64 Raspberry Pi 3 port

### PR DESCRIPTION
**Pull request by antonio-nino-diaz-arm**
This port can be compiled to boot an AArch64 or AArch32 payload with the build option `RPI3_BL33_AARCH32`.

Note: This is not a secure port of the Trusted Firmware. This port is only meant to be a reference implementation to experiment with an inexpensive board in real hardware.

This port is based on the work of Ard Biesheuvel (@ardbiesheuvel), that can be found in http://www.workofard.com/2017/02/uefi-on-the-pi/, see repository in https://github.com/ardbiesheuvel/arm-trusted-firmware/tree/rpi3. That port has been heavily modified. For starters, that port didn't manage to boot Linux or to release the secondary CPUs from their wait loop because the RAM wasn't initialized to 0 before entering the loop.

### CODE REVIEW

File: plat/rpi3/rpi3_topology.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "rpi3_private.h"
+
+/* The power domain tree descriptor */
+static unsigned char power_domain_tree_desc[] = {
+	/* Number of root nodes */
+	PLATFORM_CLUSTER_COUNT,
+	/* Number of children for the first node */
+	PLATFORM_CLUSTER0_CORE_COUNT,
+	/* Number of children for the second node */
+	PLATFORM_CLUSTER1_CORE_COUNT,
```
Comment (danh-arm) [inline plat/rpi3/rpi3_topology.c]: minor: Bit strange to define this when there's only 1 cluster.

File: docs/plat/rpi3.rst [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+IMPORTANT NOTE: This port isn't secure. All of the memory used is DRAM, which is
+available from both the Non-secure and Secure worlds. This port shouldn't be
+considered more than a prototype to play with and implement elements like PSCI
+to support the Linux kernel.
+
+Design
+------
+
+The SoC used by the Raspberry Pi 3 is the Broadcom BCM2837. It is a strange SoC
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Please remove the word "strange" - someone might take offence.

File: docs/plat/rpi3.rst [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+=======================================
+
+.. section-numbering::
+    :suffix: .
+
+.. contents::
+
+The Raspberry Pi 3 is an inexpensive single-board computer that contains four
+Cortex-A53 cores, which makes it possible to have a port of the Arm Trusted
+Firmware.
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Would be good to have a reference to the official website (e.g. https://www.raspberrypi.org/products/raspberry-pi-3-model-b/).

File: docs/plat/rpi3.rst [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+- If a file called ``kernel8.img`` is located on the ``boot`` partition of the
+  SD card, it will load it and execute in EL2 in AArch64. Basically, it executes
+  a `default AArch64 stub`_ at address ``0x0``.
+
+- If there is also a file called ``armstub8.bin``, it will load it at address
+  ``0x0`` (instead of the default one) and execute it in EL3 in AArch64. All the
+  cores are powered on at the same time and start at address ``0x0``.
+
+This is ideal for us, as we can save the default AArch32 kernel in
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Why an AArch32 kernel? Presumably both AArch64 and AArch32 can be supported.

File: plat/rpi3/platform.mk [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+ERRATA_A53_826319		:= 1
+ERRATA_A53_835769		:= 1
+ERRATA_A53_836870		:= 1
+ERRATA_A53_843419		:= 1
+ERRATA_A53_855873		:= 1
+
+# Disable the PSCI platform compatibility layer by default
+ENABLE_PLAT_COMPAT		:= 0
+
+# BL1 is overwriten by the 32-bit kernel
```
Comment (danh-arm) [inline plat/rpi3/platform.mk]: Why 32-bit kernel?

File: docs/plat/rpi3.rst [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+``kernel8.img`` and the Trusted Firmware and anything else we need in
+``armstub8.bin``, and forget about the default bootstrap code.
+
+However, we still want to load the kernel and have all cores available, which
+means that we need to make the secondary cores work in the way the kernel
+expects, as explained in `Secondary cores`_.
+
+Ideally, we want to boot the kernel in Hypervisor mode in AArch32. This means
+that BL33 can't be in EL2 in AArch64 mode. The architecture specifies that
+AArch32 Hypervisor mode isn't present when AArch64 is used for EL2.
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: This doesn't seem relevant. This port seems to support either AArch32 or AArch64 depending on the flag RPI3_BL33_IN_AARCH32.

File: docs/plat/rpi3.rst [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+~~~~~~~~~~~~~~~~~~~
+
+The file ``armstub8.bin`` contains BL1 and the FIP. It is needed to add padding
+between them so that the addresses they are loaded to match the ones specified
+when compiling the Trusted Firmware.
+
+The device tree block is loaded by the VideoCore loader from an appropriate
+file, but we can specify the address it is loaded to in ``config.txt``.
+
+The file ``kernel8.img`` contains an AArch32 kernel image, that is loaded to the
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Again, why AArch32

File: docs/plat/rpi3.rst [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+Trusted Firmware, there are 880 MiB available for the kernel.
+
+Boot sequence
+~~~~~~~~~~~~~
+
+The boot sequence of the Trusted Firmware is the usual one. The only difference
+is BL33, which is booted in AArch32 Hypervisor mode so that it can jump to the
+kernel in the same mode and let it take over that privilege level. If BL33 was
+running in EL2 in AArch64, it could only jump to the kernel in AArch32 in
+Supervisor mode.
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Or the kernel at AArch64 EL2 presumably.

File: docs/plat/rpi3.rst [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    CPU register settings:
+      r0 = 0
+      r1 = Machine type number (3138 for the Rasbperry Pi 3)
+      r2 = Physical address of the device tree block (dtb) in RAM
+
+    CPU mode:
+      All forms of interrupts must be disabled (IRQs and FIQs)
+
+The AArch32 bootstrap takes care of the setup of the registers, and BL2 setups
+the BL33 context so that all interrupts are disabled.
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: This is specific to the AArch32 kernel. The ARM64 kernel has a different boot interface.

File: docs/plat/rpi3.rst [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+Secondary cores
+~~~~~~~~~~~~~~~
+
+The 32-bit kernel used by Raspbian doesn't have support for PSCI, so it is
```
Comment (danh-arm) [inline docs/plat/rpi3.rst]: Ah - so now I see why all the AArch32 references. Is there not an AArch64 kernel available for RPi3? If not, this doc should say somewhere why an AArch32 kernel is specifically referenced, while also saying AArch64 kernels are theoretically supported by this port.

File: docs/plat/rpi3.rst (hunk 3 above)
Comment (antonio-nino-diaz-arm) [reply docs/plat/rpi3.rst]: Done.

File: docs/plat/rpi3.rst (hunk 2 above)
Comment (antonio-nino-diaz-arm) [reply docs/plat/rpi3.rst]: Done.

File: docs/plat/rpi3.rst (hunk 4 above)
Comment (antonio-nino-diaz-arm) [reply docs/plat/rpi3.rst]: I've added a short paragraph at the start of the documentation explaining the references to the AArch32 kernel. Also, the instructions to have a AArch64 kernel aren't that complicated, and most of this documentation applies to both kernels, so I've modified a bit the whole document in order to explain how to use both.

File: docs/plat/rpi3.rst (hunk 9 above)
Comment (antonio-nino-diaz-arm) [reply docs/plat/rpi3.rst]: I've removed this part and just kept the reference to the file with the information.

File: plat/rpi3/platform.mk (hunk 5 above)
Comment (antonio-nino-diaz-arm) [reply plat/rpi3/platform.mk]: I've changed this comment to a generic one and have the documentation explain the reason why this option is used.

File: plat/rpi3/rpi3_topology.c (hunk 1 above)
Comment (antonio-nino-diaz-arm) [reply plat/rpi3/rpi3_topology.c]: Removed.

**Comment by danh-arm** [comment]
jenkins: test this please

**Pull request merged**